tests/runner.sh

```diff
@@ -1,16 +1,29 @@
 #!/usr/bin/env bash
 
-for i in test_*.sh; do
-    echo -ne "$i    "
-    out=${i%.*}.out
-    ok=${i%.*}.ok
-    diff=${i%.*}.diff
-    ./$i > $out 2>&1
-    diff $out $ok > $diff
-    if [ -s $diff ]; then
-        echo "fail"
-        exit 1
-    else
-        echo "pass"
-    fi
-done
+# runner.sh main  (only main tests)
+# runner.sh extra (only extra tests)
+
+run_tests() {
+    for i in "$@"; do
+        echo -ne "$i    "
+        out=${i%.*}.out
+        ok=${i%.*}.ok
+        diff=${i%.*}.diff
+        ./$i > $out 2>&1
+        diff $out $ok > $diff
+        if [ -s $diff ]; then
+            echo "fail"
+            exit 1
+        else
+            echo "pass"
+        fi
+    done
+}
+
+if [ "$1" == "main" ]; then
+    run_tests $(ls test_*.sh)
+elif [ "$1" == "extra" ]; then
+    run_tests $(ls extra/test_*.sh)
+else
+    echo "Invalid argument: $1"
+fi
```

Comment:
Not good. See other comment.

Just changing the for to use the provided arguments makes the function a lot more flexible.